src/cli/MareTF.cpp

```diff
@@ -181,6 +185,273 @@ template<> tferr_t& tferr_t::operator<<<tfendl_t>(const tfendl_t&) {
 
 } // namespace
 
+namespace DistanceMapping {
+
+enum Leaf : uint8_t {
+	IN,         // leaf is opaque
+	OUT,        // leaf is transparent
+	FIGUREITOUT // leaf is mixed but small so you should brute force scan.
+};
+
+enum Quadrant : uint8_t {
+	NW,
+	NE,
+	SW,
+	SE
+};
+
+static uint16_t i32tou16sat(int32_t i) {
+	if (i > UINT16_MAX)
+		return UINT16_MAX;
+	if (i < 0)
+		return 0;
+	return (uint16_t) i;
+}
+static uint16_t ftoabsu16(float f) { return i32tou16sat((int32_t) fabs(f)); }
+
+class QuadTree {
+public:
+	QuadTree(
+		std::vector<std::byte> *actuallyrgba32323232f,
+		uint16_t inWidth,
+		uint16_t inHeight,
+		uint16_t reduceX,
+		uint16_t reduceY,
+		float distanceSpread,
+		bool sampleCentered,
+		float alphaThreshold = 0.04f
+	)
+		: rgba32323232f(reinterpret_cast<const float *>(actuallyrgba32323232f->data()))
+		, imgWidth(inWidth)
+		, imgHeight(inHeight)
+		, reduceX(reduceX)
+		, reduceY(reduceY)
+		, searchRadius(ftoabsu16(2.0f * std::max((float) this->reduceX, (float) this->reduceY) * distanceSpread))
+		// the second power of two after search radius
+		, granularity(((uint16_t) 4) << ftoabsu16(ceil(log2((float) searchRadius))))
+		, offsX(sampleCentered ? this->reduceX / 2 : 0)
+		, offsY(sampleCentered ? this->reduceY / 2 : 0)
+		, alphaThreshold(alphaThreshold)
+		, root([this](){ return scanImg(0, 0, imgWidth, imgHeight); }())
+	{}
+	~QuadTree() { Node::wither(root); }
+
+private:
+	const float *rgba32323232f;
+public:
+	const uint16_t imgWidth, imgHeight;
+	const uint16_t reduceX, reduceY;
+	const uint16_t searchRadius;
+private:
+	const uint16_t granularity;
+public:
+	const uint16_t offsX, offsY;
+private:
+	const float alphaThreshold;
+public:
+
+	class Node;
+	typedef std::variant<Leaf, Node *> Branch;
+	const Branch root;
+
+	class Node {
+	public:
+		Branch branches[4];
+		Node(Branch nw, Branch ne, Branch sw, Branch se) : branches {nw, ne, sw, se} {};
+		~Node() {
+			for (uint16_t q = NW; q <= SE; q++)
```

Comment:
```suggestion
			for (uint16_t q = 0; q <= this->branches.size(); q++)
```

`std::array::size` is constexpr, this is fine. Also I use `this->` to refer to member methods and fields at all times